File: Stack/typescript/stack_doubly_linked_list.ts
import * as readline from 'readline'
const rl = readline.createInterface({
    input:process.stdin
})
let input_data:string = ""
rl.on("line",(line)=>{
    input_data+=line
    if(line.includes("-1")){
        rl.close()
    }
})

class Node<T>{
    prev: Node<T> | null
    val : T
    next: Node<T> | null
    constructor(val:T){
        this.prev = null
        this.val  = val
        this.next = null
    }
}

class Stack<T>{
    sz:number
    head:Node<T> | null
    tail:Node<T> | null
    constructor(){
        this.sz = 0
        this.head = null
        this.tail = null
    }

    top():T|undefined{
        return this.tail?.val
    }

    pop():void{
        if (this.tail === null){
            return
        }
        this.tail = this.tail.prev
        if (this.tail!=null){
            this.tail.next = null
        }
        this.sz--
    }
    push(val:T):void{
        const newNode = new Node(val)
        if(this.head===null){
            this.head = newNode
            this.tail = newNode
            this.sz++
            return
        }
        if(this.tail){
            this.tail.next = newNode
            newNode.prev = this.tail
            this.tail = newNode
            this.sz++
        }
    }

    size():number{
        return this.sz
    }

    isEmpty():boolean{
        return this.sz==0
    }
}

const stack = new Stack<number>()
rl.on("close",()=>{
    const a = input_data.split(" ").map(Number)
    let i = 0;
    while(true){
        if(a[i]===-1){
            break
        }else{
            stack.push(a[i])
        }
        i++
    }
    while(!stack.isEmpty()){
        process.stdout.write(stack.top() +  " ")
        stack.pop()
    }
    console.log()
})

File: Stack/python/stack_doubly_linked_list.py
import sys
class Node:
    def __init__(self,val):
        self.prev = None
        self.val  = val
        self.next = None

class Stack:
    def __init__(self):
        self.sz = -1
        self.head = None
        self.tail = None
    
    def top(self):
        return self.tail.val if self.tail != None else None

    def pop(self):
        if self.tail == None:
            return
        delete_node = self.tail
        self.tail = self.tail.prev
        if self.tail != None:
            self.tail.next = None
        del delete_node
        self.sz-=1
    
    def push(self,val):
        newNode = Node(val)
        if self.head == None:
            self.head = newNode
            self.tail = newNode
            self.sz+=1
            return
        self.tail.next = newNode
        newNode.prev = self.tail
        self.tail = newNode
        self.sz+=1

    def size(self):
        return self.sz
    
    def isEmpty(self):
        return self.sz==-1

input_data = sys.stdin.read().strip().split(" ")
a = list(map(int,input_data))
stack = Stack()
i = 0
while True:
    if a[i]==-1:
        break
    else:
        stack.push(a[i])
    i+=1
while not stack.isEmpty():
    print(stack.top(),end=" ")
    stack.pop()




